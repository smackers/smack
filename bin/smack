#!/usr/bin/env python3
#
# This file is distributed under the MIT License. See LICENSE for details.
#

import os
import sys

sys.dont_write_bytecode = True # prevent creation of .pyc files
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../share'))

import smack.top
smack.top.main()
