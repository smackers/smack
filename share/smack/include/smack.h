//
// This file is distributed under the MIT License. See LICENSE for details.
//
#ifndef SMACK_H_
#define SMACK_H_
#include <limits.h>

/**
 * The SMACK "prelude" declarations
 */

#ifdef __cplusplus
extern "C" {
#endif

// Apprently needed by SVCOMP benchmarks
#define __inline
#define __builtin_expect __builtinx_expect
#define __builtin_memcpy __builtinx_memcpy
#define __builtin_va_start __builtinx_va_start
#define __builtin_object_size __builtinx_object_size

// For handling of va_start macro
void __builtinx_va_start(char*,char*);

void __SMACK_code(const char *fmt, ...);
void __SMACK_mod(const char *fmt, ...);
void __SMACK_decl(const char *fmt, ...);
void __SMACK_top_decl(const char *fmt, ...);

typedef struct smack_value { void* dummy; }* smack_value_t;
smack_value_t __SMACK_value();
smack_value_t __SMACK_values(void* ary, unsigned count);
smack_value_t __SMACK_return_value(void);

// Sugar for __SMACK_init_func_XXX()
#define __SMACK_INIT(x) void __SMACK_init_func_##x()

#if MEMORY_SAFETY
// Inserts memory access checks in form of assert to check null pointer access
// and buffer overflow errors
void __SMACK_check_memory_safety(void*, unsigned long);
void __SMACK_check_memory_leak();
#endif

// We need this to enforce that assert/assume are function calls
// with an integer argument (DSA gets confused otherwise)
__attribute__((always_inline)) void __SMACK_dummy(int v);

void __VERIFIER_assume(int);
#ifndef CUSTOM_VERIFIER_ASSERT
void __VERIFIER_assert(int);
#endif
void __VERIFIER_error(void);
void exit(int);

#ifndef AVOID_NAME_CONFLICTS
#define assert(EX) __VERIFIER_assert(EX)
#define assume(EX) __VERIFIER_assume(EX)
#endif

#define S4(a,b,c,d) a b c d
#define S3(a,b,c) a b c
#define S2(a,b) a b
#define S1(a) a
#define U4(a,b,c,d) a ## _ ## b ## _ ## c ## _ ## d
#define U3(a,b,c) a ## _ ## b ## _ ## c
#define U2(a,b) a ## _ ## b
#define U1(a) a

#define TY(_1, _2, _3, _4, A, ...) A

#define S(...) TY(__VA_ARGS__, S4, S3, S2, S1)(__VA_ARGS__)
#define U(...) TY(__VA_ARGS__, U4, U3, U2, U1)(__VA_ARGS__)

#define NONDET_DECL(P, ty...) S(ty) U(P,U(ty)) ()

void* __VERIFIER_nondet(void);
NONDET_DECL(__SMACK_nondet,char);
NONDET_DECL(__SMACK_nondet,signed,char);
NONDET_DECL(__SMACK_nondet,unsigned,char);
NONDET_DECL(__SMACK_nondet,short);
NONDET_DECL(__SMACK_nondet,short,int);
NONDET_DECL(__SMACK_nondet,signed,short);
NONDET_DECL(__SMACK_nondet,signed,short,int);
NONDET_DECL(__SMACK_nondet,unsigned,short);
NONDET_DECL(__SMACK_nondet,unsigned,short,int);
NONDET_DECL(__SMACK_nondet,int);
NONDET_DECL(__SMACK_nondet,signed,int);
NONDET_DECL(__SMACK_nondet,unsigned);
NONDET_DECL(__SMACK_nondet,unsigned,int);
NONDET_DECL(__SMACK_nondet,long);
NONDET_DECL(__SMACK_nondet,long,int);
NONDET_DECL(__SMACK_nondet,signed,long);
NONDET_DECL(__SMACK_nondet,signed,long,int);
NONDET_DECL(__SMACK_nondet,unsigned,long);
NONDET_DECL(__SMACK_nondet,unsigned,long,int);
NONDET_DECL(__SMACK_nondet,long,long);
NONDET_DECL(__SMACK_nondet,long,long,int);
NONDET_DECL(__SMACK_nondet,signed,long,long);
NONDET_DECL(__SMACK_nondet,signed,long,long,int);
NONDET_DECL(__SMACK_nondet,unsigned,long,long);
NONDET_DECL(__SMACK_nondet,unsigned,long,long,int);
NONDET_DECL(__VERIFIER_nondet,char);
NONDET_DECL(__VERIFIER_nondet,signed,char);
NONDET_DECL(__VERIFIER_nondet,unsigned,char);
NONDET_DECL(__VERIFIER_nondet,short);
NONDET_DECL(__VERIFIER_nondet,short,int);
NONDET_DECL(__VERIFIER_nondet,signed,short);
NONDET_DECL(__VERIFIER_nondet,signed,short,int);
NONDET_DECL(__VERIFIER_nondet,unsigned,short);
NONDET_DECL(__VERIFIER_nondet,unsigned,short,int);
NONDET_DECL(__VERIFIER_nondet,int);
NONDET_DECL(__VERIFIER_nondet,signed,int);
NONDET_DECL(__VERIFIER_nondet,unsigned);
NONDET_DECL(__VERIFIER_nondet,unsigned,int);
NONDET_DECL(__VERIFIER_nondet,long);
NONDET_DECL(__VERIFIER_nondet,long,int);
NONDET_DECL(__VERIFIER_nondet,signed,long);
NONDET_DECL(__VERIFIER_nondet,signed,long,int);
NONDET_DECL(__VERIFIER_nondet,unsigned,long);
NONDET_DECL(__VERIFIER_nondet,unsigned,long,int);
NONDET_DECL(__VERIFIER_nondet,long,long);
NONDET_DECL(__VERIFIER_nondet,long,long,int);
NONDET_DECL(__VERIFIER_nondet,signed,long,long);
NONDET_DECL(__VERIFIER_nondet,signed,long,long,int);
NONDET_DECL(__VERIFIER_nondet,unsigned,long,long);
NONDET_DECL(__VERIFIER_nondet,unsigned,long,long,int);
NONDET_DECL(__VERIFIER_nondet,float);
NONDET_DECL(__VERIFIER_nondet,double);
NONDET_DECL(__VERIFIER_nondet,long,double);

#if FLOAT_ENABLED
//floats
float fabsf(float x);
float fdimf(float x, float y);
float roundf(float x);
//The following 3 functions are incomplete pending rounding mode implementation
float rintf(float x);
float nearbyintf(float x);
long lrintf(float x);
long lroundf(float x);
float floorf(float x);
float ceilf(float x);
float truncf(float x);
float sqrtf(float x);
float remainderf(float x, float y);
float fminf(float x, float y);
float fmaxf(float x, float y);
float fmodf(float x, float y);
float modff(float x, float* y);
float copysignf(float x, float y);
int __isnormalf(float x);
int __isSubnormalf(float x);
int __iszerof(float x);
int __isinff(float x);
int __isnanf(float x);
int __isnegativef(float x);
int __ispositivef(float x);
int __signbitf(float x);
int signbitf(float x);
int __fpclassifyf(float x);
int fpclassifyf(float x);
int __finitef(float x);
//float nan(float x);

//doubles
double fabs(double x);
double fdim(double x, double y);
double round(double x);
long lround(double x);
//The following 3 functions are incomplete pending rounding mode implementation
double rint(double x);
double nearbyint(double x);
long lrint(double x);
double floor(double x);
double ceil(double x);
double trunc(double x);
double sqrt(double x);
double remainder(double x, double y);
double fmin(double x, double y);
double fmax(double x, double y);
double fmod(double x, double y);
double modf(double x, double* y);
double copysign(double x, double y);
double nan(const char* x);
int __isnormal(double x);
int __isSubnormal(double x);
int __iszero(double x);
int __isinf(double x);
int __isnan(double x);
int __isnegative(double x);
int __ispositive(double x);
int __signbit(double x);
int signbit(double x);
int __fpclassify(double x);
int fpclassify(double x);
int __finite(double x);

//long doubles
/*int __isnormall(long double x);
int __isSubnormall(long double x);
int __iszerol(long double x);
int __isinfl(long double x);
int __isnanl(long double x);
int __isnegativel(long double x);
int __ispositivel(long double x);
int __signbitl(long double x);
int signbitl(long double x);
int __fpclassifyl(long double x);
int fpclassifyl(long double x);
int __finitel(long double x);*/
#endif

#undef S1
#undef S2
#undef S3
#undef S4
#undef U1
#undef U2
#undef U3
#undef U4
#undef TY
#undef S
#undef U
#undef NONDET_DECL

// Apparently used in SVCOMP benchmarks
_Bool __VERIFIER_nondet_bool(void);
unsigned char __VERIFIER_nondet_uchar(void);
unsigned short __VERIFIER_nondet_ushort(void);
unsigned __VERIFIER_nondet_uint(void);
unsigned long __VERIFIER_nondet_ulong(void);
void* __VERIFIER_nondet_pointer(void);


void __SMACK_decls();

#ifdef __cplusplus
}
#endif

#endif /*SMACK_H_*/
