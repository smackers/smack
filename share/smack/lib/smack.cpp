#include <cstddef>
#include <new>

extern void *malloc(size_t size);
extern void free(void *p);

void *operator new(size_t size) throw(std::bad_alloc) { return malloc(size); }

void operator delete(void *p) throw() { free(p); }
