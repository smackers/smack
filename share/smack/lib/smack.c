//
// This file is distributed under the MIT License. See LICENSE for details.
//
#include <limits.h>
#include <smack.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>

/**
 * The SMACK "prelude" definitions
 *
 * TODO add more documentation
 *
 * NOTES ON MEMORY MODELS
 *
 * 1. addresses are (unbounded) integers
 * 2. one unbounded integer is stored at each address
 * 3. (NO-REUSE only) heap addresses are allocated in a strictly increasing
 *    fashion
 * 4. (NO-REUSE only) freed (heap) addresses are never reallocated
 * 5. the address space is partitioned as follows
 *
 * Address A                                        Allocation
 * ---------                                        ----------
 * A > 0                                            Heap
 * A = 0                                            Not allocated (null)
 * $GLOBALS_BOTTOM <= A < 0                         Static (global storage)
 * $GLOBALS_BOTTOM - 32768 <= A < $GLOBALS_BOTTOM   Not allocated (padding)
 * $EXTERNS_BOTTOM <= A < $GLOBALS_BOTTOM - 32768   External globals
 * A < $EXTERNS_BOTTOM                              Objects returned from
 * external functions
 *
 */

void __VERIFIER_assume(int x) {
  __SMACK_dummy(x);
  __SMACK_code("assume @ != $0;", x);
}

#ifndef CUSTOM_VERIFIER_ASSERT
void __VERIFIER_assert(int x) {
#if !MEMORY_SAFETY && !SIGNED_INTEGER_OVERFLOW_CHECK
  __SMACK_dummy(x);
  __SMACK_code("assert @ != $0;", x);
#endif
}
#endif

void __VERIFIER_error(void) {
#if !MEMORY_SAFETY && !SIGNED_INTEGER_OVERFLOW_CHECK
  __SMACK_code("assert false;");
#elif MEMORY_SAFETY
  __SMACK_code("assert {:valid_memtrack} $allocatedCounter == 0;");
  __SMACK_code("assume false;");
#else
  __SMACK_code("assume false;");
#endif
}

void __SMACK_check_overflow(int flag) {
  __SMACK_dummy(flag);
  __SMACK_code("assert {:overflow} @ == $0;", flag);
}

char __VERIFIER_nondet_char(void) {
  char x = __SMACK_nondet_char();
  __VERIFIER_assume(x >= SCHAR_MIN && x <= SCHAR_MAX);
  return x;
}

signed char __VERIFIER_nondet_signed_char(void) {
  signed char x = __SMACK_nondet_signed_char();
  __VERIFIER_assume(x >= SCHAR_MIN && x <= SCHAR_MAX);
  return x;
}

unsigned char __VERIFIER_nondet_unsigned_char(void) {
  unsigned char x = __SMACK_nondet_unsigned_char();
  __VERIFIER_assume(x >= 0 && x <= UCHAR_MAX);
  return x;
}

short __VERIFIER_nondet_short(void) {
  short x = __SMACK_nondet_short();
  __VERIFIER_assume(x >= SHRT_MIN && x <= SHRT_MAX);
  return x;
}

signed short __VERIFIER_nondet_signed_short(void) {
  signed short x = __SMACK_nondet_signed_short();
  __VERIFIER_assume(x >= SHRT_MIN && x <= SHRT_MAX);
  return x;
}

signed short int __VERIFIER_nondet_signed_short_int(void) {
  signed short int x = __SMACK_nondet_signed_short_int();
  __VERIFIER_assume(x >= SHRT_MIN && x <= SHRT_MAX);
  return x;
}

unsigned short __VERIFIER_nondet_unsigned_short(void) {
  unsigned short x = __SMACK_nondet_unsigned_short();
  __VERIFIER_assume(x >= 0 && x <= USHRT_MAX);
  return x;
}

unsigned short int __VERIFIER_nondet_unsigned_short_int(void) {
  unsigned short int x = __SMACK_nondet_unsigned_short_int();
  __VERIFIER_assume(x >= 0 && x <= USHRT_MAX);
  return x;
}

int __VERIFIER_nondet_int(void) {
  int x = __SMACK_nondet_int();
  __VERIFIER_assume(x >= INT_MIN && x <= INT_MAX);
  return x;
}

signed int __VERIFIER_nondet_signed_int(void) {
  signed int x = __SMACK_nondet_signed_int();
  __VERIFIER_assume(x >= INT_MIN && x <= INT_MAX);
  return x;
}

unsigned __VERIFIER_nondet_unsigned(void) {
  unsigned x = __SMACK_nondet_unsigned();
  unsigned min = __SMACK_nondet_unsigned();
  unsigned max = __SMACK_nondet_unsigned();
  __VERIFIER_assume(min == 0 && max >= UINT_MAX && max <= UINT_MAX);
  __VERIFIER_assume(x >= min && x <= max);
  return x;
}

unsigned int __VERIFIER_nondet_unsigned_int(void) {
  unsigned int x = __SMACK_nondet_unsigned_int();
  unsigned int min = __SMACK_nondet_unsigned_int();
  unsigned int max = __SMACK_nondet_unsigned_int();
  __VERIFIER_assume(min == 0 && max >= UINT_MAX && max <= UINT_MAX);
  __VERIFIER_assume(x >= min && x <= max);
  return x;
}

long __VERIFIER_nondet_long(void) {
  long x = __SMACK_nondet_long();
  __VERIFIER_assume(x >= LONG_MIN && x <= LONG_MAX);
  return x;
}

long int __VERIFIER_nondet_long_int(void) {
  long int x = __SMACK_nondet_long_int();
  __VERIFIER_assume(x >= LONG_MIN && x <= LONG_MAX);
  return x;
}

signed long __VERIFIER_nondet_signed_long(void) {
  signed long x = __SMACK_nondet_signed_long();
  __VERIFIER_assume(x >= LONG_MIN && x <= LONG_MAX);
  return x;
}

signed long int __VERIFIER_nondet_signed_long_int(void) {
  signed long int x = __SMACK_nondet_signed_long_int();
  __VERIFIER_assume(x >= LONG_MIN && x <= LONG_MAX);
  return x;
}

unsigned long __VERIFIER_nondet_unsigned_long(void) {
  unsigned long x = __SMACK_nondet_unsigned_long();
  unsigned long min = __SMACK_nondet_unsigned_long();
  unsigned long max = __SMACK_nondet_unsigned_long();
  __VERIFIER_assume(min == 0 && max >= ULONG_MAX && max <= ULONG_MAX);
  __VERIFIER_assume(x >= min && x <= max);
  return x;
}

unsigned long int __VERIFIER_nondet_unsigned_long_int(void) {
  unsigned long int x = __SMACK_nondet_unsigned_long_int();
  unsigned long int min = __SMACK_nondet_unsigned_long_int();
  unsigned long int max = __SMACK_nondet_unsigned_long_int();
  __VERIFIER_assume(min == 0 && max >= ULONG_MAX && max <= ULONG_MAX);
  __VERIFIER_assume(x >= min && x <= max);
  return x;
}

long long __VERIFIER_nondet_long_long(void) {
  long long x = __SMACK_nondet_long_long();
  __VERIFIER_assume(x >= LLONG_MIN && x <= LLONG_MAX);
  return x;
}

long long int __VERIFIER_nondet_long_long_int(void) {
  long long int x = __SMACK_nondet_long_long_int();
  __VERIFIER_assume(x >= LLONG_MIN && x <= LLONG_MAX);
  return x;
}

signed long long __VERIFIER_nondet_signed_long_long(void) {
  signed long long x = __SMACK_nondet_signed_long_long();
  __VERIFIER_assume(x >= LLONG_MIN && x <= LLONG_MAX);
  return x;
}

signed long long int __VERIFIER_nondet_signed_long_long_int(void) {
  signed long long int x = __SMACK_nondet_signed_long_long_int();
  __VERIFIER_assume(x >= LLONG_MIN && x <= LLONG_MAX);
  return x;
}

unsigned long long __VERIFIER_nondet_unsigned_long_long(void) {
  unsigned long long x = __SMACK_nondet_unsigned_long_long();
  unsigned long long min = __SMACK_nondet_unsigned_long_long();
  unsigned long long max = __SMACK_nondet_unsigned_long_long();
  __VERIFIER_assume(min == 0 && max >= ULLONG_MAX && max <= ULLONG_MAX);
  __VERIFIER_assume(x >= min && x <= max);
  return x;
}

unsigned long long int __VERIFIER_nondet_unsigned_long_long_int(void) {
  unsigned long long int x = __SMACK_nondet_unsigned_long_long_int();
  unsigned long long int min = __SMACK_nondet_unsigned_long_long_int();
  unsigned long long int max = __SMACK_nondet_unsigned_long_long_int();
  __VERIFIER_assume(min == 0 && max >= ULLONG_MAX && max <= ULLONG_MAX);
  __VERIFIER_assume(x >= min && x <= max);
  return x;
}

// Used in SVCCOMP benchmarks

_Bool __VERIFIER_nondet_bool(void) {
  _Bool x = (_Bool)__VERIFIER_nondet_int();
  __VERIFIER_assume(x == 0 || x == 1);
  return x;
}

unsigned char __VERIFIER_nondet_uchar(void) {
  unsigned char x = __VERIFIER_nondet_unsigned_char();
  return x;
}

unsigned short __VERIFIER_nondet_ushort(void) {
  unsigned short x = __VERIFIER_nondet_unsigned_short();
  return x;
}

unsigned int __VERIFIER_nondet_uint(void) {
  unsigned int x = __VERIFIER_nondet_unsigned_int();
  return x;
}

unsigned long __VERIFIER_nondet_ulong(void) {
  unsigned long x = __VERIFIER_nondet_unsigned_long();
  return x;
}

void *__VERIFIER_nondet_pointer(void) { return __VERIFIER_nondet(); }

void __SMACK_dummy(int v) { __SMACK_code("assume true;"); }

#define D(d) __SMACK_top_decl(d)

void __SMACK_decls(void) {
  // Memory debugging symbols
  D("type $mop;");
  D("procedure boogie_si_record_mop(m: $mop);");
  D("const $MOP: $mop;");

  D("var $exn: bool;");
  D("var $exnv: int;");

  D("procedure $alloc(n: ref) returns (p: ref)\n"
    "{\n"
    "  call p := $$alloc(n);\n"
    "}\n");

#if MEMORY_SAFETY
  __SMACK_dummy((int)__SMACK_check_memory_safety);
  D("implementation __SMACK_check_memory_safety(p: ref, size: ref)\n"
    "{\n"
    "  assert {:valid_deref} $Alloc[$base(p)];\n"
    "  assert {:valid_deref} $sle.ref.bool($base(p), p);\n"
#if MEMORY_MODEL_NO_REUSE_IMPLS
    "  assert {:valid_deref} $sle.ref.bool($add.ref(p, size), "
    "$add.ref($base(p), $Size($base(p))));\n"
#elif MEMORY_MODEL_REUSE
    "  assert {:valid_deref} $sle.ref.bool($add.ref(p, size), "
    "$add.ref($base(p), $Size[$base(p)]));\n"
#else
    "  assert {:valid_deref} $sle.ref.bool($add.ref(p, size), "
    "$add.ref($base(p), $Size($base(p))));\n"
#endif
    "}\n");

  D("function $base(ref) returns (ref);");
  D("var $allocatedCounter: int;\n");


#if MEMORY_MODEL_NO_REUSE_IMPLS
  D("var $Alloc: [ref] bool;");
  D("function $Size(ref) returns (ref);");
  D("var $CurrAddr:ref;\n");

  D("procedure $malloc(n: ref) returns (p: ref)\n"
    "modifies $allocatedCounter, $Alloc, $CurrAddr;\n"
    "{\n"
    "  if ($ne.ref.bool(n, $0.ref)) {\n"
    "    $allocatedCounter := $allocatedCounter + 1;\n"
    "  }\n"
    "  call p := $$alloc(n);\n"
    "}\n");

  D("procedure $galloc(base_addr: ref, size: ref)\n"
    "{\n"
    "  assume $Size(base_addr) == size;\n"
    "  assume (forall addr: ref :: {$base(addr)} $sle.ref.bool(base_addr, "
    "addr) && $slt.ref.bool(addr, $add.ref(base_addr, size)) ==> $base(addr) "
    "== base_addr);\n"
    "  $Alloc[base_addr] := true;\n"
    "}\n");

  D("procedure {:inline 1} $$alloc(n: ref) returns (p: ref)\n"
    "modifies $Alloc, $CurrAddr;\n"
    "{\n"
    "  assume $sle.ref.bool($0.ref, n);\n"
    "  if ($slt.ref.bool($0.ref, n)) {\n"
    "    p := $CurrAddr;\n"
    "    havoc $CurrAddr;\n"
    "    assume $sge.ref.bool($sub.ref($CurrAddr, n), p);\n"
    "    assume $sgt.ref.bool($CurrAddr, $0.ref) && $slt.ref.bool($CurrAddr, "
    "$MALLOC_TOP);\n"
    "    assume $Size(p) == n;\n"
    "    assume (forall q: ref :: {$base(q)} $sle.ref.bool(p, q) && "
    "$slt.ref.bool(q, $add.ref(p, n)) ==> $base(q) == p);\n"
    "    $Alloc[p] := true;\n"
    "  } else {\n"
    "    p := $0.ref;\n"
    "  }\n"
    "}\n");

  D("procedure $free(p: ref)\n"
    "modifies $Alloc, $allocatedCounter;\n"
    "{\n"
    "  if ($ne.ref.bool(p, $0.ref)) {\n"
    "    assert {:valid_free} $eq.ref.bool($base(p), p);\n"
    "    assert {:valid_free} $Alloc[p];\n"
    "    assert {:valid_free} $slt.ref.bool($0.ref, p);\n"
    "    $Alloc[p] := false;\n"
    "    $allocatedCounter := $allocatedCounter - 1;\n"
    "  }\n"
    "}\n");

#elif MEMORY_MODEL_REUSE // can reuse previously-allocated and freed addresses
  D("var $Alloc: [ref] bool;");
  D("var $Size: [ref] ref;\n");

  D("procedure $malloc(n: ref) returns (p: ref)\n"
    "modifies $allocatedCounter, $Alloc, $Size;\n"
    "{\n"
    "  if ($ne.ref.bool(n, $0.ref)) {\n"
    "    $allocatedCounter := $allocatedCounter + 1;\n"
    "  }\n"
    "  call p := $$alloc(n);\n"
    "}\n");


  D("procedure $galloc(base_addr: ref, size: ref);\n"
    "modifies $Alloc, $Size;\n"
    "ensures $Size[base_addr] == size;\n"
    "ensures (forall addr: ref :: {$base(addr)} $sle.ref.bool(base_addr, addr) "
    "&& $slt.ref.bool(addr, $add.ref(base_addr, size)) ==> $base(addr) == "
    "base_addr);\n"
    "ensures $Alloc[base_addr];\n"
    "ensures (forall q: ref :: {$Size[q]} q != base_addr ==> $Size[q] == "
    "old($Size[q]));\n"
    "ensures (forall q: ref :: {$Alloc[q]} q != base_addr ==> $Alloc[q] == "
    "old($Alloc[q]));\n");

  D("procedure {:inline 1} $$alloc(n: ref) returns (p: ref);\n"
    "modifies $Alloc, $Size;\n"
    "ensures $sle.ref.bool($0.ref, n);\n"
    "ensures $slt.ref.bool($0.ref, n) ==> $slt.ref.bool($0.ref, p) && "
    "$slt.ref.bool(p, $sub.ref($MALLOC_TOP, n));\n"
    "ensures $eq.ref.bool(n, $0.ref) ==> p == $0.ref;\n"
    "ensures !old($Alloc[p]);\n"
    "ensures (forall q: ref :: old($Alloc[q]) ==> ($slt.ref.bool($add.ref(p, "
    "n), q) || $slt.ref.bool($add.ref(q, $Size[q]), p)));\n"
    "ensures $Alloc[p];\n"
    "ensures $Size[p] == n;\n"
    "ensures (forall q: ref :: {$Size[q]} q != p ==> $Size[q] == "
    "old($Size[q]));\n"
    "ensures (forall q: ref :: {$Alloc[q]} q != p ==> $Alloc[q] == "
    "old($Alloc[q]));\n"
    "ensures (forall q: ref :: {$base(q)} $sle.ref.bool(p, q) && "
    "$slt.ref.bool(q, $add.ref(p, n)) ==> $base(q) == p);\n");

  D("procedure $free(p: ref);\n"
    "modifies $Alloc, $allocatedCounter;\n"
    "requires $eq.ref.bool(p, $0.ref) || ($eq.ref.bool($base(p), p) && "
    "$Alloc[p]);\n"
    "requires $slt.ref.bool($0.ref, p);\n"
    "ensures $ne.ref.bool(p, $0.ref) ==> !$Alloc[p];\n"
    "ensures $ne.ref.bool(p, $0.ref) ==> (forall q: ref :: {$Alloc[q]} q != p "
    "==> $Alloc[q] == old($Alloc[q]));\n"
    "ensures $ne.ref.bool(p, $0.ref) ==> $allocatedCounter == "
    "old($allocatedCounter) - 1;\n");

#else // NO_REUSE does not reuse previously-allocated addresses
  D("var $Alloc: [ref] bool;");
  D("function $Size(ref) returns (ref);");
  D("var $CurrAddr:ref;\n");

  D("procedure $malloc(n: ref) returns (p: ref)\n"
    "modifies $allocatedCounter, $Alloc, $CurrAddr;\n"
    "{\n"
    "  if ($ne.ref.bool(n, $0.ref)) {\n"
    "    $allocatedCounter := $allocatedCounter + 1;\n"
    "  }\n"
    "  call p := $$alloc(n);\n"
    "}\n");


  D("procedure $galloc(base_addr: ref, size: ref);\n"
    "modifies $Alloc;\n"
    "ensures $Size(base_addr) == size;\n"
    "ensures (forall addr: ref :: {$base(addr)} $sle.ref.bool(base_addr, addr) "
    "&& $slt.ref.bool(addr, $add.ref(base_addr, size)) ==> $base(addr) == "
    "base_addr);\n"
    "ensures $Alloc[base_addr];\n"
    "ensures (forall q: ref :: {$Alloc[q]} q != base_addr ==> $Alloc[q] == "
    "old($Alloc[q]));\n");

  D("procedure {:inline 1} $$alloc(n: ref) returns (p: ref);\n"
    "modifies $Alloc, $CurrAddr;\n"
    "ensures $sle.ref.bool($0.ref, n);\n"
    "ensures $slt.ref.bool($0.ref, n) ==> $sge.ref.bool($sub.ref($CurrAddr, "
    "n), old($CurrAddr)) && p == old($CurrAddr);\n"
    "ensures $sgt.ref.bool($CurrAddr, $0.ref) && $slt.ref.bool($CurrAddr, "
    "$MALLOC_TOP);\n"
    "ensures $slt.ref.bool($0.ref, n) ==> $Size(p) == n;\n"
    "ensures $slt.ref.bool($0.ref, n) ==> (forall q: ref :: {$base(q)} "
    "$sle.ref.bool(p, q) && $slt.ref.bool(q, $add.ref(p, n)) ==> $base(q) == "
    "p);\n"
    "ensures $slt.ref.bool($0.ref, n) ==> $Alloc[p];\n"
    "ensures $eq.ref.bool(n, $0.ref) ==> old($CurrAddr) == $CurrAddr && p == "
    "$0.ref;\n"
    "ensures $eq.ref.bool(n, $0.ref)==> $Alloc[p] == old($Alloc)[p];\n"
    "ensures (forall q: ref :: {$Alloc[q]} q != p ==> $Alloc[q] == "
    "old($Alloc[q]));\n");

  D("procedure $free(p: ref);\n"
    "modifies $Alloc, $allocatedCounter;\n"
    "requires $eq.ref.bool(p, $0.ref) || ($eq.ref.bool($base(p), p) && "
    "$Alloc[p]);\n"
    "requires $slt.ref.bool($0.ref, p);\n"
    "ensures $ne.ref.bool(p, $0.ref) ==> !$Alloc[p];\n"
    "ensures $ne.ref.bool(p, $0.ref) ==> (forall q: ref :: {$Alloc[q]} q != p "
    "==> $Alloc[q] == old($Alloc[q]));\n"
    "ensures $ne.ref.bool(p, $0.ref) ==> $allocatedCounter == "
    "old($allocatedCounter) - 1;\n");
#endif

#else

#if MEMORY_MODEL_NO_REUSE_IMPLS
  D("var $CurrAddr:ref;\n");

  D("procedure $malloc(n: ref) returns (p: ref)\n"
    "modifies $CurrAddr;\n"
    "{\n"
    "  call p := $$alloc(n);\n"
    "}\n");

  D("procedure {:inline 1} $$alloc(n: ref) returns (p: ref)\n"
    "modifies $CurrAddr;\n"
    "{\n"
    "  assume $sge.ref.bool(n, $0.ref);\n"
    "  if ($sgt.ref.bool(n, $0.ref)) {\n"
    "    p := $CurrAddr;\n"
    "    havoc $CurrAddr;\n"
    "    assume $sge.ref.bool($sub.ref($CurrAddr, n), p);\n"
    "    assume $sgt.ref.bool($CurrAddr, $0.ref) && $slt.ref.bool($CurrAddr, "
    "$MALLOC_TOP);\n"
    "  } else {\n"
    "    p := $0.ref;\n"
    "  }\n"
    "}\n");

  D("procedure $free(p: ref);\n");

#elif MEMORY_MODEL_REUSE // can reuse previously-allocated and freed addresses
  D("var $Alloc: [ref] bool;");
  D("var $Size: [ref] ref;\n");

  D("procedure $malloc(n: ref) returns (p: ref)\n"
    "modifies $Alloc, $Size;\n"
    "{\n"
    "  call p := $$alloc(n);\n"
    "}\n");

  D("procedure {:inline 1} $$alloc(n: ref) returns (p: ref);\n"
    "modifies $Alloc, $Size;\n"
    "ensures $sle.ref.bool($0.ref, n);\n"
    "ensures $slt.ref.bool($0.ref, n) ==> $slt.ref.bool($0.ref, p) && "
    "$slt.ref.bool(p, $sub.ref($MALLOC_TOP, n));\n"
    "ensures $eq.ref.bool(n, $0.ref) ==> p == $0.ref;\n"
    "ensures !old($Alloc[p]);\n"
    "ensures (forall q: ref :: old($Alloc[q]) ==> ($slt.ref.bool($add.ref(p, "
    "n), q) || $slt.ref.bool($add.ref(q, $Size[q]), p)));\n"
    "ensures $Alloc[p];\n"
    "ensures $Size[p] == n;\n"
    "ensures (forall q: ref :: {$Size[q]} q != p ==> $Size[q] == "
    "old($Size[q]));\n"
    "ensures (forall q: ref :: {$Alloc[q]} q != p ==> $Alloc[q] == "
    "old($Alloc[q]));\n");

  D("procedure $free(p: ref);\n"
    "modifies $Alloc;\n"
    "ensures !$Alloc[p];\n"
    "ensures (forall q: ref :: {$Alloc[q]} q != p ==> $Alloc[q] == "
    "old($Alloc[q]));\n");

#else // NO_REUSE does not reuse previously-allocated addresses
  D("var $CurrAddr:ref;\n");

  D("procedure $malloc(n: ref) returns (p: ref)\n"
    "modifies $CurrAddr;\n"
    "{\n"
    "  call p := $$alloc(n);\n"
    "}\n");

  D("procedure {:inline 1} $$alloc(n: ref) returns (p: ref);\n"
    "modifies $CurrAddr;\n"
    "ensures $sle.ref.bool($0.ref, n);\n"
    "ensures $slt.ref.bool($0.ref, n) ==> $sge.ref.bool($sub.ref($CurrAddr, "
    "n), old($CurrAddr)) && p == old($CurrAddr);\n"
    "ensures $sgt.ref.bool($CurrAddr, $0.ref) && $slt.ref.bool($CurrAddr, "
    "$MALLOC_TOP);\n"
    "ensures $eq.ref.bool(n, $0.ref) ==> old($CurrAddr) == $CurrAddr && p == "
    "$0.ref;\n");

  D("procedure $free(p: ref);\n");
#endif
#endif

#undef D
}

#if MEMORY_SAFETY
void __SMACK_check_memory_leak(void) {
  __SMACK_code("assert {:valid_memtrack} $allocatedCounter == 0;");
}
#endif

void __SMACK_init_func_memory_model(void) {
#if MEMORY_MODEL_NO_REUSE || MEMORY_MODEL_NO_REUSE_IMPLS
  __SMACK_code("$CurrAddr := $1024.ref;");
#endif
#if MEMORY_SAFETY
  __SMACK_code("$allocatedCounter := 0;");
#endif
}
