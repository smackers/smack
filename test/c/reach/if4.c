#include "smack.h"

int main() {
  int x;
  x = 3;
  if (x == 4)
    x = 5;
  x = 4;

  return x;
}
