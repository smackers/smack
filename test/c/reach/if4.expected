{"/home/mcarter/projects/smack-project/smack/install/include/smack/smack.h": [[40, 26], [41, 3], [42, 1], [44, 26], [45, 3], [46, 1], [57, 3], [58, 3], [59, 3]], "if4.c": [[6, 14]]}
